### ISSUE TITLE
Proxy commands from global.json/project.json

### ISSUE THREAD

**Issue by Mpdreamz** [dotnet/cli#910]
I'm new to the dotnet-cli tooling so please forgive any ignorance on this post 

It would be nice if we could use the `dotnet` at the root of a checkout too. 

Other then [putting executable on $PATH](https://github.com/dotnet/cli/blob/master/Documentation/intro-to-cli.md#adding-a-new-command-locally) it would be nice if we could override/add commands in global.json:

```json
{
   "projects" : [ "src" ],
   "commands" : {
        "compile" : { "win" : "build.cmd build", "nix" : "build.sh build" },
        "test" : { 
            "xplat" : "packages/build/Fake/Fake.exe test",
            "description" : "compiles and runs all the unit tests"
        }
   }
}
```

After checkout I could simply call `dotnet compile` in the root on any platform and it would compile all the projects. As suppose to each project rolling their own build.[cmd|ps1|sh|etcetera] and documenting which to call on what platform. Even nicer the experience for command discovery could be ubiquitous 

```
user ~/rootofproject
$ dotnet commands

global commands:
- compile
└─ compiles and runs all the unit tests
- test

project commands:

all: 
- compile
- restore
- other commands on $path

src/mvcapp/project.json:
- mycrazycommand defined in "commands" in project.json
```

This would rely on two new commands 

* `commands` 
* `listprojects` taking `global.config`'s `"projects"` settings into account spits out all the project.json files. Something that e.g omnisharp could also use to standardize project discovery.

another feature request would be to override defaults when calling commands on project.json files

```json
{
   "projects" : [ "src" ],
   "commands" : {
        "compile" : { "win" : "build.cmd build", "nix" : "build.sh build" },
        "test" : { "win" : "build.cmd test", "nix" : "build.sh test" },
   },
   "projectCommands": {
       "restore" : { "win" : { }, "nix" : "" }
   }
}
```

this would proxy e.g the restore command on any project.json to e.g Paket.

**Comment by davidfowl** [comment]
The CLI supports adding a tools section to projects that expose commands that can be run with dotnet {blah}. For e.g.

```JSON
{
    ...
    "tools": {
      "dotnet-sometool": "1.0.0-*"
    }
}
```

You can navigate to the project folder and go `dotnet sometool` and have that work. Similarly, you can replace inbox commands this way (I believe @piotrpMSFT correct me if I'm wrong).

The natural expansion would be to allow tools in `global.json` that apply to all projects. I'm not sure about overriding project commands there though, that would require some more thought.

Also running paket via *still* means it would need to generate a `project.lock.json` file.

**Comment by Mpdreamz** [comment]
minus paket, will open that can of worms else where :smile: 

I would love if `tools` can be renamed to `commands` (wasn't it commands in DNX?) and it supported more then just calling tools on the PATH.

I realise "tools" are our `devDependencies` but they should not double as our `scripts` too.

```
"tools" : {
    "dotnet-sometool" : "1.0.0-*", //pulls in (dotnet-)sometool
},
"commands" : {
    "sometool" : "", //declares a (dotnet-)tool on $path
    "mycommand" : "SomeAssembly args", //oldschool dnx finding a main to run in SomeAssemly 
    "mycommand2: {
        "any" : "proxy commandline to execute on any platform",
        "win" : "proxy commandline to execute on windows ",
        "nix" : "proxy commandline to execute on *Nix",
        "tool" :"sometool", //longform
        "run" : { "assembly" : "MyAssembly", "args" : "x* }, //longform
        "description" : "meta"
    }
}
```

I'm looking for an ubiquitous command line experience when cloning a .NET repos. E.g in node land I can clone any repos and call npm from the root e.g `npm install` to pull down any deps. 
Node repos are 9/10 times single project repositories though where as .NET are multi-project/solution based. Having to navigate anywhere post clone kinda sucks :smile:

**Comment by piotrpMSFT** [comment]
The team is actively working on enabling [MSBuild](https://blogs.msdn.microsoft.com/dotnet/2016/05/23/changes-to-project-json/) and the component affected will be superseded by the new project system, so I am closing this issue.

Of course once that happens... custom targets and tasks and Execs will unblock this whole set of scenarios :)
